activerecord/lib/active_record/errors.rb

```diff
@@ -123,6 +123,28 @@ class RecordNotUnique < WrappedDatabaseException
   class InvalidForeignKey < WrappedDatabaseException
   end
 
+  # Raised when a foreign key constraint cannot be added because the column type does not match the referenced column type.
+  class MismatchedForeignKey < WrappedDatabaseException
+    def initialize(message)
+      parts = message.scan(/`(\w+)`[ $)]/).flatten
```

Comment:
I'm uncomfortable about teaching a generic-sounding exception (admittedly, currently only used by one in-tree adapter) how to parse a particular DBMS's error message... even before potential i18n gets involved.